### ISSUE TITLE
Running gem5 networking stack without dpdk

### ISSUE THREAD

**Issue by sammy17** [architecture-research-group/gem5-dpdk-setup#1]
Hi, thank you for the great work! 
I want to run the kernel network stack (without DPDK) with your setup. 
Are there already scripts available to do this?
Also, have you already done this? If so, I am interested in knowing the maximum achievable network throughput without DPDK.  

If not, could you please give me an idea of how this can be done? 
Here, I am assuming your hardware changes are compatible with the kernel network stack as well, please correct me if I am wrong.

Thank you in advance.

**Comment by jumeike** [comment]
Hi Sammy,

Yes, you can use the kernel network stack and we have scripts (e.g. [memcached-kernel.sh](https://github.com/architecture-research-group/gem5-dpdk-setup/blob/main/run-scripts/memcached-kernel.sh), [memcached_kernel.sh](https://github.com/architecture-research-group/gem5-dpdk-setup/blob/main/guest-scripts/memcached_kernel.sh), etc.) to do this when running memcachedKernel (for instance). Please check [this](https://github.com/architecture-research-group/gem5-dpdk-setup/wiki/How-to-run-memcached-on-gem5) out to see how to do this with our EtherLoadGen model. Otherwise, you can also use [dual mode gem5](https://github.com/architecture-research-group/gem5-dpdk-setup/wiki/Dual-Mode-gem5-Simulation).

Yes, we have done this using iperf and achieves a maximum bandwidth of 8.9Gbps with an OoO core running at 4GHz. Please also check [this](https://github.com/architecture-research-group/gem5-dpdk-setup/wiki/Dual-Mode-gem5-Simulation#step-3) out.

You can also modify the disk image to add other network-intensive applications and run over the kernel network stack in gem5. See [this](https://github.com/architecture-research-group/gem5-dpdk-setup/wiki/Dual-Mode-gem5-Simulation#step-2) to learn see how to do this.

Please let us know if this helps. Thank you.

**Comment by sammy17** [comment]
Thank you very much for the quick response and information @jumeike !
I will try running the dual mode and EtherLoadGen. 

Related to the maximum achievable bandwidth without DPDK, I am observing approximately the same value for single-core throughput in the default gem5. 
I also tried multicore throughput (running 4 iperf3 instances in parallel with four O3 ARM cores running at 3GHz in gem5-dist) and the throughput did not scale well with the number of cores. It was capped at around 12 Gbps. 

Do you have any idea what might be the bottleneck in this case? 
I noticed the DDIO enhancement that you did in gem5, I am currently trying it out. 

I know my questions are not directly related to DPDK and I really appreciate your response.

**Comment by jumeike** [comment]
Hi @sammy17,

I am not sure what this happened with dist-gem5. There is a need to synchronize the different instances of gem5 when using dist-gem5. During my experiments, I had an issue with using O3 ARM cores to restore from the checkpoint when using dist-gem5. As a result I moved to dual-mode gem5.

Did you try on real-system to know how much throughput scales with the number of cores? There may be some interesting insights.

As discussed with you via email, the DDIO enhancements are available in this repo
